src/net_processing.cpp

```diff
@@ -1734,16 +1734,27 @@ void static ProcessGetData(CNode& pfrom, const CChainParams& chainparams, CConnm
             connman.PushMessage(&pfrom, msgMaker.Make(nSendFlags, NetMsgType::TX, *tx));
             mempool.RemoveUnbroadcastTx(tx->GetHash());
             // As we're going to send tx, make sure its unconfirmed parents are made requestable.
-            for (const auto& txin : tx->vin) {
-                auto txinfo = mempool.info(txin.prevout.hash);
-                if (txinfo.tx && txinfo.m_time > now - UNCONDITIONAL_RELAY_DELAY) {
-                    // Relaying a transaction with a recent but unconfirmed parent.
-                    if (WITH_LOCK(pfrom.m_tx_relay->cs_tx_inventory, return !pfrom.m_tx_relay->filterInventoryKnown.contains(txin.prevout.hash))) {
-                        LOCK(cs_main);
-                        State(pfrom.GetId())->m_recently_announced_invs.insert(txin.prevout.hash);
+            std::vector<uint256> parent_ids_to_add;
+            {
+                LOCK(mempool.cs);
+                auto txiter = mempool.GetIter(tx->GetHash());
+                if (txiter) {
+                    const CTxMemPool::setEntries& parents = mempool.GetMemPoolParents(*txiter);
+                    parent_ids_to_add.reserve(parents.size());
+                    for (CTxMemPool::txiter parent_iter : parents) {
+                        if (parent_iter->GetTime() > now - UNCONDITIONAL_RELAY_DELAY) {
+                            parent_ids_to_add.push_back(parent_iter->GetTx().GetHash());
+                        }
                     }
                 }
             }
+            for (const uint256& parent_txid : parent_ids_to_add) {
+                // Relaying a transaction with a recent but unconfirmed parent.
+                if (WITH_LOCK(pfrom.m_tx_relay->cs_tx_inventory, return !pfrom.m_tx_relay->filterInventoryKnown.contains(parent_txid))) {
+                    LOCK(cs_main);
```

Comment:
It's probably fine in practice, and it's not a regression in this PR, but locking and unlocking cs_tx_inventory and cs_main (potentially) for each parent kind of bugs me each time I look at it, so I had a go at trying to avoid it. Came up with:

```c++
{
    LOCK(pfrom.m_tx_relay->cs_tx_inventory);
    auto is_known = [&](uint256& parent_txid) EXCLUSIVE_LOCKS_REQUIRED(pfrom.m_tx_relay->cs_tx_inventory) { return pfrom.m_tx_relay->filterInventoryKnown.contains(parent_txid); };
    parent_ids_to_add.erase(std::remove_if(parent_ids_to_add.begin(), parent_ids_to_add.end(), is_known), parent_ids_to_add.end());
}
if (!parent_ids_to_add.empty()) {
    LOCK(cs_main);
    for (const uint256& parent_txid : parent_ids_to_add) {
        // Relaying a transaction with a recent but unconfirmed parent.
        State(pfrom.GetId())->m_recently_announced_invs.insert(parent_txid);
    }
}
```

Reply:
Calling `State()` to iterate over every peer for every input txid also seems wasteful, especially since we're already in a loop over all getdata items. I think better would be:

```
            if (!parent_ids_to_add.empty()) {
                LOCK(cs_main);
                CNodeState* node = State(pfrom.GetId());
                for (const uint256& parent_txid : parent_ids_to_add) {
                    // Relaying a transaction with a recent but unconfirmed parent.
                    node->m_recently_announced_invs.insert(parent_txid);
                }
            }
```

I also don't understand the benefit of adding the clang lock annotation to the lambda. It's only in scope here, so it's never going to be called from somewhere else without `pfrom.m_tx_relay->cs_tx_inventory` being held.

All of this locking/unlocking of multiple mutexes goes away once peer data is consolidated into a single `Peer` object.

Reply:
I started to do this, but this doesn't exactly add to the readability of the code; can we punt on this to some future refactor, especially if we think existing refactors under way would change all this anyway?  

In practice there should be more than 25 unconfirmed parents anyway, and I wouldn't expect much lock contention either.

Reply:
Definitely seems fine to punt it to me

Reply:
@jnewbery Accessing guarded data structures in a lambda means the lambda needs to be annotated with the required lock (or an AssertLockHeld or LockAssertion added cf #19668), or there will be compile time thread safety errors. Doesn't matter how limited the lambda's scope is.